src/aiida/tools/dumping/processes.py

```diff
@@ -0,0 +1,448 @@
+    # ? should also create one here, if the function is imported normally and used in Python scripts
+    output_path = validate_make_dump_path(path=output_path, overwrite=overwrite)
+
+    # This will eventually be replaced once pydantic backend PR merged
+    if node_dumper is None:
+        node_dumper = ProcessNodeYamlDumper()
+    node_dumper.dump_yaml(process_node=process_node, output_path=output_path.resolve())
+
+    # Don't increment index for `ProcessNodes` that don't (always?) have file IO
+    # (`CalcFunctionNodes`/`WorkFunctionNodes`), such as `create_kpoints_from_distance`
+    called_links = process_node.base.links.get_outgoing(link_type=(LinkType.CALL_CALC, LinkType.CALL_WORK)).all()
+    called_links = [
+        called_link
+        for called_link in called_links
+        if not isinstance(called_link.node, (CalcFunctionNode, WorkFunctionNode))
+    ]
+
+    for index, link_triple in enumerate(sorted(called_links, key=lambda link_triple: link_triple.node.ctime), start=1):
+        child_node = link_triple.node
+        child_label = generate_node_input_label(index=index, link_triple=link_triple)
+
+        output_path_child = output_path.resolve() / child_label
+
+        # Recursive function call for `WorkChainNode``
+        if isinstance(child_node, WorkChainNode):
+            process_dumped = workchain_dump(
+                process_node=child_node,
+                output_path=output_path_child,
+                no_node_inputs=no_node_inputs,
+                use_presubmit=use_presubmit,
+                node_dumper=node_dumper,
+            )
+
+        # Dump for `CalcJobNode`
+        elif isinstance(child_node, CalcJobNode):
+            process_dumped = calcjob_dump(
+                calcjob_node=child_node,
+                output_path=output_path_child,
+                no_node_inputs=no_node_inputs,
+                use_presubmit=use_presubmit,
+                node_dumper=node_dumper,
+            )
+
+        else:
+            process_dumped = False
+
+        all_processes_dumped = all_processes_dumped and process_dumped
+
+    return all_processes_dumped
+
+
+# Separate functions for CalcJob dumping using pre_submit, as well as for the node_inputs
+def calcjob_node_inputs_dump(calcjob_node: CalcJobNode, output_path: Path, inputs_relpath: Path = Path('node_inputs')):
+    """Dump inputs of a `CalcJobNode` of type `SinglefileData` and `FolderData`.
+
+    :param calcjob_node: The `CalcJobNode` whose inputs will be dumped.
+    :type calcjob_node: CalcJobNode
+    :param output_path: The path where the inputs will be dumped.
+    :type output_path: Path
+    """
+    dump_types = (SinglefileData, FolderData)
+
+    # ? Not using the `node_class` argument of `get_incoming`, as it does not actually retrieve, e.g. a `UpfData` node
+    # ? (due to planned deprecation?)
+    # ? Instead, check for isinstance of `SinglefileData`
```

Comment:
The reason this doesn't work is because the class hierarchy matching in the querybuilder is not based upon the Python class hierarchy but an artificial hierarchy in the entry points strings. For example, if we have the following entry points
```
'core.singlefile' = 'aiida.orm.nodes.data.singlefile:SinglefileData'`
'core.upf' = 'aiida.orm.nodes.data.upf:UpfData'
'core.singlefile.hypothetical_upf' = 'aiida.orm.nodes.data.hypothetical:HypotheticalUpfData'
```
where `UpfData` and `HypotheticalUpfData` are both a sublcass of `SinglefileData`. Only `HypotheticalUpfData` would be matched as subclasses of `SinglefileData` in queryies because the entry point string `core.singlefile.hypothetical_upf` starts with the entry point `core.singlefile` of the base class.

Reply:
Ah, alright, I see. Makes sense! Thanks for the clarification :)